contribute/design/1.x/ndm/20190705-ndm-disk-identification.md

```diff
@@ -0,0 +1,214 @@
+* [Summary](#summary)
+* [Motivation](#motivation)
+    * [Goals](#goals)
+    * [Non-Goals](#non-goals)
+* [Proposal](#proposal)
+    * [User Stories(#user-stories)
+      * [Identify a physical disk](#identify-a-physical-disk)
+      * [Identify a virtual disk](#identify-a-virtual-disk)
+    * [Implementation Details/Notes/Constraints](#implementation-detailsnotesconstraints)
+      * [Current Implementation](#current-implementation)
+      * [Shortcomings of current implementation](#shortcomings-of-current-implementation)
+      * [Proposed Implementation](#proposed-implementation)
+      * [Workflow](#workflow)
+* [Graduation Criteria](#graduation-criteria)
+* [Implementation History](#implementation-history)
+* [Drawbacks](#drawbacks)
+* [Infrastructure Needed](#infrastructure-needed)
+
+## Summary
+
+This proposal brings out the design details to implement a Disk identification
+solution which can be used to uniquely identify a disk across the cluster.
+
+## Motivation
+
+Uniquely identifying a disk across a cluster is a major challenge which needs to 
+be tackled. The identification is necessary, due to the fact that disk can move
+from node to node and even attached to different ports on the same node. The unique
+identification solution will help the consumers of NDM to track the disk
+and thus the data movement within the cluster.
+
+### Goals
+
+- A unique disk identification mechanism that will work across the cluster on
+  disks having atleast a GPT label
+- Should be able to identify same disk attached at multiple paths.
+- Should be able to generate separate IDs for partitions
+
+### Non-Goals
+
+- Identification of disk attached to 2 nodes at the same time
+
+## Proposal
+
+### User Stories
+
+#### Identify a physical disk
+NDM should be able to generate a unqiue id for a physical disk attached to the node
+and this id should not change even if the disk is attached at a different port on the
+same host or moved to a different node or altogether moved to a different cluster.
+
+#### Identify a virtual disk
+NDM should be able to generate a unique id for Virtual Disks(which may not have all 
+relevant information in VPD). This id should also persist between restarts of the node,
+attaching to a different machine or a different SCSI port of the same machine.
+
+### Implementation Details/Notes/Constraints
+
+The implementation details for the disk identification. The current implementation 
+works well for physical disks which have a unique WWN and model assigned by the 
+manufacturer. But the implementation fails for `Virtual_disk`, because the fields
+are provider dependent and are not guaranteed to be unique. 
+
+#### Current Implementation
+The following data is fetched from the disk: 
+1. `ID_WWN`
+2. `ID_MODEL`
+3. `ID_SERIAL_SHORT`
+4. `ID_VENDOR`
+5. `ID_TYPE`
+
+If the `ID_MODEL` is not `Virtual_disk` or `EphemeralDisk`:
+	- field 1-4 are appended
+	- md5 hash of the result is generated which will be UID for the disk by NDM
```

Comment:
What if the backing LUN is a shared resource and attached to 2 nodes.?

Reply:
In case of shared resource, like a single disk attached to 2 nodes. This design will not work? Because in the resulting object that we store in `etcd` will have 2 hostname and nodename for the same disk.